--- ppp-2.4.4/pppdump/Makefile.linux	2013-02-17 15:01:34.263566117 +0800
+++ ppp-2.4.4new/pppdump/Makefile.linux	2013-02-17 15:18:16.409913846 +0800
@@ -11,7 +11,7 @@
 all:	pppdump
 
 pppdump: $(OBJS)
-	$(CC) -o pppdump $(OBJS)
+	$(CC) $(LDFLAGS) -o pppdump $(OBJS)
 
 clean:
 	rm -f pppdump $(OBJS) *~
--- ppp-2.4.4/pppd/plugins/radius/Makefile.linux	2013-02-17 15:01:34.263566117 +0800
+++ ppp-2.4.4new/pppd/plugins/radius/Makefile.linux	2013-02-17 15:14:36.268371263 +0800
@@ -44,13 +44,13 @@
 	$(INSTALL) -c -m 444 pppd-radattr.8 $(MANDIR)
 
 radius.so: radius.o libradiusclient.a
-	$(CC) -o radius.so -shared radius.o libradiusclient.a
+	$(CC) $(LDFLAGS) -o radius.so -shared radius.o libradiusclient.a
 
 radattr.so: radattr.o
-	$(CC) -o radattr.so -shared radattr.o
+	$(CC) $(LDFLAGS) -o radattr.so -shared radattr.o
 
 radrealms.so: radrealms.o
-	$(CC) -o radrealms.so -shared radrealms.o
+	$(CC) $(LDFLAGS) -o radrealms.so -shared radrealms.o
 
 CLIENTOBJS = avpair.o buildreq.o config.o dict.o ip_util.o \
 	clientid.o sendserver.o lock.o util.o md5.o
--- ppp-2.4.4/pppd/plugins/Makefile.linux	2006-06-04 13:04:14.000000000 +0800
+++ ppp-2.4.4new/pppd/plugins/Makefile.linux	2013-02-17 15:10:53.745977938 +0800
@@ -1,7 +1,5 @@
-CC	= gcc
 COPTS	= -O2 -g
 CFLAGS	= $(COPTS) -I.. -I../../include -fPIC
-LDFLAGS	= -shared
 INSTALL	= install
 
 DESTDIR = $(INSTROOT)@DESTDIR@
@@ -23,7 +21,7 @@
 	for d in $(SUBDIRS); do $(MAKE) $(MFLAGS) -C $$d all; done
 
 %.so: %.c
-	$(CC) -o $@ $(LDFLAGS) $(CFLAGS) $^
+	$(CC) -o $@ $(LDFLAGS) -shared $(CFLAGS) $^
 
 VERSION = $(shell awk -F '"' '/VERSION/ { print $$2; }' ../patchlevel.h)
 
--- ppp-2.4.4/pppd/plugins/pppoatm/Makefile.linux	2004-11-14 09:12:10.000000000 +0800
+++ ppp-2.4.4new/pppd/plugins/pppoatm/Makefile.linux	2013-02-17 15:23:43.907389881 +0800
@@ -1,7 +1,5 @@
-CC	= gcc
 COPTS	= -O2 -g
 CFLAGS	= $(COPTS) -I../.. -I../../../include -fPIC
-LDFLAGS	= -shared
 INSTALL	= install
 
 #***********************************************************************
@@ -33,7 +31,7 @@
 all: $(PLUGIN)
 
 $(PLUGIN): $(PLUGIN_OBJS)
-	$(CC) $(CFLAGS) -o $@ -shared $^ $(LIBS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ -shared $^ $(LIBS)
 
 install: all
 	$(INSTALL) -d -m 755 $(LIBDIR)
--- ppp-2.4.4/pppd/plugins/rp-pppoe/Makefile.linux	2013-02-17 15:01:34.215566166 +0800
+++ ppp-2.4.4new/pppd/plugins/rp-pppoe/Makefile.linux	2013-02-17 15:17:00.649529825 +0800
@@ -29,13 +29,13 @@
 all: rp-pppoe.so pppoe-discovery
 
 pppoe-discovery: pppoe-discovery.o utils.o libplugin.a
-	$(CC) -o pppoe-discovery pppoe-discovery.o utils.o libplugin.a
+	$(CC) $(LDFLAGS) -o pppoe-discovery pppoe-discovery.o utils.o libplugin.a
 
 pppoe-discovery.o: pppoe-discovery.c
 	$(CC) $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o pppoe-discovery.o pppoe-discovery.c
 
 rp-pppoe.so: libplugin.a plugin.o
-	$(CC) -o rp-pppoe.so -shared plugin.o libplugin.a
+	$(CC) $(LDFLAGS) -o rp-pppoe.so -shared plugin.o libplugin.a
 
 install: all
 	$(INSTALL) -d -m 755 $(LIBDIR)
--- ppp-2.4.4/pppstats/Makefile.linux	2013-02-17 15:01:34.215566166 +0800
+++ ppp-2.4.4new/pppstats/Makefile.linux	2013-02-17 15:19:28.850264840 +0800
@@ -26,7 +26,7 @@
 	$(INSTALL) -c -m 444 pppstats.8 $(MANDIR)
 
 pppstats: $(PPPSTATSRCS)
-	$(CC) $(CFLAGS) -o pppstats pppstats.c $(LIBS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o pppstats pppstats.c $(LIBS)
 
 clean:
 	rm -f pppstats *~ #* core
--- ppp-2.4.4/chat/Makefile.linux	2013-02-17 15:01:34.215566166 +0800
+++ ppp-2.4.4new/chat/Makefile.linux	2013-02-17 15:09:27.865112363 +0800
@@ -18,7 +18,7 @@
 all:	chat
 
 chat:	chat.o
-	$(CC) -o chat chat.o
+	$(CC) $(LDFLAGS) -o chat chat.o
 
 chat.o:	chat.c
 	$(CC) -c $(CFLAGS) -o chat.o chat.c
