--- lua-5.1.4/src/loslib.c	2013-02-28 14:05:26.658112828 +0800
+++ lua-5.1.4new/src/loslib.c	2013-02-28 15:39:51.228035873 +0800
@@ -129,8 +129,17 @@
     stm = gmtime(&t);
     s++;  /* skip `!' */
   }
-  else
+  else {
+    FILE *fp;
+    char buf[128], zone[128];
+    fp=popen("uci get system.@system[0].timezone", "r");
+    fgets(buf, sizeof(buf), fp);
+    pclose(fp);
+    sprintf(zone, "TZ=%s", buf);
+    putenv(zone);
+    tzset();
     stm = localtime(&t);
+  }
   if (stm == NULL)  /* invalid date? */
     lua_pushnil(L);
   else if (strcmp(s, "*t") == 0) {
